minidump-processor/src/stackwalker/arm64.rs

```diff
@@ -205,88 +257,62 @@ where
+    // several pointers when they get pushed to memory, including the return
+    // address (lr) and frame pointer (fp), which both get pushed at the start
+    // of most non-leaf functions.
     //
-    // Rather than actually thinking about how to recover the key and properly
-    // decode this, let's apply a simple heuristic. We get the maximum address
-    // that's contained in a module we know about, which will have some highest
-    // bit that is set. We can then safely mask out any bit that's higher than
-    // that one, which will hopefully mask out all the weird security stuff
-    // in the high bits.
-    if let Some(last_module) = modules.by_addr().next_back() {
-        // Get the highest mappable address
-        let mut mask = last_module.base_address() + last_module.size();
-        // Repeatedly OR this value with its shifted self to "smear" its
-        // highest set bit down to all lower bits. This will get us a
-        // mask we can use to AND out any bits that are higher.
-        mask |= mask >> 1;
-        mask |= mask >> 1;
-        mask |= mask >> 2;
-        mask |= mask >> 4;
-        mask |= mask >> 8;
-        mask |= mask >> 16;
-        mask |= mask >> 32;
-        let stripped = ptr & mask;
-
-        // Only actually use this stripped value if it ended up pointing in
-        // a module so we don't start corrupting normal pointers that are just
-        // in modules we don't know about.
-        if modules.module_at_address(stripped).is_some() {
-            // trace!("stripped pointer {:016x} -> {:016x}", ptr, stripped);
-            return stripped;
-        }
-    }
-
-    ptr
+    // We lack some of the proper context to implement the "strip" primitive, because
+    // the amount of bits that are "real" pointer depends on various extensions like
+    // pointer tagging and how big page tables are. If we allocate too many bits to
+    // "real" then we can get ptr_auth bits in our pointers, and if we allocate too
+    // few we can end up truncating our pointers. Thankfully we'll usually have a bit
+    // a bit of margin from pointers not having the highest real bits set.
+    //
+    // To help us guess, we have a few pieces of information:
+    //
+    // * Apple seems to default to a 24/40 split, so 40 bits for the "real" is a good baseline
+    // * We know the address ranges of various loaded/unloaded modules
+    // * We know the address range of the stacks
+    // * We *can* know the address range of some sections of the heap (MemoryList)
+    // * We *can* know the page mappings (MemoryInfo)
+    //
+    // Right now we only incorporate the first two. Ideally we would process all those sources
+    // once at the start of stack walking and pass it down to the ARM stackwalker but that's
+    // a lot of annoying rewiring that won't necessarily improve results.
+    let apple_default_max_addr = (1 << 40) - 1;
+    let max_module_addr = modules
+        .by_addr()
+        .next_back()
```

Comment:
I agree it can/should be precomputed but I'm not sure what the shape of the API should be. My changes have however effectively *optimized* this code, and the inputs are currently all relatively cheap/constants. For now I'd like to leave it uncached to leave room for us using more intelligent/subtle information in the future.

That said I wouldn't reject a PR that hoisted this stuff up to process_minidump and passed it down as part of some more principled/convenient args struct or something..?

Reply:
Maybe framehop's solution can be of inspiration: It precomputes the mask. https://docs.rs/framehop/latest/framehop/aarch64/struct.PtrAuthMask.html